src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py

```diff
@@ -82,11 +82,187 @@ def test_paralooptrans_validate_force(fortran_reader):
+
+
+def test_paralooptrans_collapse_options(fortran_reader, fortran_writer):
+    '''
+    Test the 'collapse' option, also in combination with the 'force' and
+    'ignore_dependencies_for' options.
+    '''
+    psyir = fortran_reader.psyir_from_source('''
+        subroutine my_sub()
+          integer :: i, j, k
+          real :: var(10,10,10) = 1
+          integer, dimension(10) :: map
+
+          do i = 1, 10  ! This loop is iteration independent
+            do j = 1, 10  ! This loop has a loop-carried dependency in var1
+              do k = 1, j  ! This loop bounds depend on previous indices
+                var(i,j,k) = var(i, map(j), k)
+              end do
+            end do
+          end do
+        end subroutine my_sub''')
+    loop = psyir.walk(Loop)[0]
+    trans = ParaTrans()
+    # The validation does not see any problem because the outer loop can be
+    # parallel
+    trans.validate(loop)
+
+    # By default it stops at collapse 1, because loop 2 has a dependency on,
+    # var. With verbose it adds the stopping reason as a comment.
+    test_loop = psyir.copy().walk(Loop, stop_type=Loop)[0]
+    trans.apply(test_loop, {"collapse": True, "verbose": True})
+    assert '''\
+!$omp parallel do collapse(1) default(shared), private(i,j,k)
+do i = 1, 10, 1
+  ! Error: The write access to 'var(i,j,k)' and the read access to 'var(i,\
+map(j),k)' are dependent and cannot be parallelised. Variable: 'var'.
+  do j = 1, 10, 1
+    do k = 1, j, 1
+      var(i,j,k) = var(i,map(j),k)
+    enddo
+  enddo
+enddo
+''' in fortran_writer(test_loop.parent.parent)
+
+    # Check that the collapse logic can uses the "ignore_dependencies_for" and
+    # "force" logic to skip dependencies
+    test_loop = psyir.copy().walk(Loop, stop_type=Loop)[0]
+    trans.apply(test_loop, {"collapse": True, "verbose": True,
+                            "ignore_dependencies_for": ["var"]})
+    assert '''\
+!$omp parallel do collapse(2) default(shared), private(i,j,k)
+do i = 1, 10, 1
+  do j = 1, 10, 1
+    ! Loop can not be collapsed because one of the bounds depends on the \
+previous iteration variable 'j'
+    do k = 1, j, 1
+      var(i,j,k) = var(i,map(j),k)
+    enddo
+  enddo
+enddo
+''' in fortran_writer(test_loop.parent.parent)
+
+    test_loop = psyir.copy().walk(Loop, stop_type=Loop)[0]
+    trans.apply(test_loop, {"collapse": True, "verbose": True, "force": True})
+    assert '''\
+!$omp parallel do collapse(2) default(shared), private(i,j,k)
+do i = 1, 10, 1
+  do j = 1, 10, 1
+    ! Loop can not be collapsed because one of the bounds depends on the \
+previous iteration variable 'j'
+    do k = 1, j, 1
+      var(i,j,k) = var(i,map(j),k)
+    enddo
+  enddo
+enddo
+''' in fortran_writer(test_loop.parent.parent)
+
+    # Also it won't collapse if the loop inside is not perfectly nested,
+    # regardless of the force option, but we consider this expected and
+    # verbose won't log the reason.
```

Comment:
We might consider it expected but a user might not, especially if the loop body is large and hard to follow. I suggest we therefore log the reason for this too.

Reply:
Ok done